### ISSUE TITLE
Invalid results passing ByVal ArrayType

### ISSUE THREAD

**Issue by maleadt** [intel/intel-graphics-compiler#251]
The following LLVM IR, translated to SPIR-V using the Khronos LLVM to SPIR-V translator, results in code that behaves incorrectly:

```llvm
define spir_kernel void @_Z17julia_kernel_59985TupleI4Int8E7LLVMPtrIS0_Li1EE([1 x i8]* nocapture nonnull readonly byval([1 x i8]) align 1 dereferenceable(1) %0, i8 addrspace(1)* %1) local_unnamed_addr #0 {
top:
  %2 = getelementptr inbounds [1 x i8], [1 x i8]* %0, i64 0, i64 0
  %3 = load i8, i8* %2, align 1
  store i8 %3, i8 addrspace(1)* %1, align 1
  ret void
}
```

```
; SPIR-V
; Version: 1.1
; Generator: Khronos LLVM/SPIR-V Translator; 14
; Bound: 18
; Schema: 0
               OpCapability Addresses
               OpCapability Kernel
               OpCapability Int64
               OpCapability Int8
          %1 = OpExtInstImport "OpenCL.std"
               OpMemoryModel Physical64 OpenCL
               OpEntryPoint Kernel %10 "_Z17julia_kernel_60665TupleI4Int8E7LLVMPtrIS0_Li1EE"
               OpSource OpenCL_C 200000
               OpName %top "top"
               OpDecorate %11 FuncParamAttr ByVal
               OpDecorate %11 FuncParamAttr NoCapture
               OpDecorate %11 FuncParamAttr NoWrite
               OpDecorate %11 Alignment 1
               OpDecorate %11 MaxByteOffset 1
      %ulong = OpTypeInt 64 0
      %uchar = OpTypeInt 8 0
    %ulong_1 = OpConstant %ulong 1
    %ulong_0 = OpConstant %ulong 0
       %void = OpTypeVoid
%_arr_uchar_ulong_1 = OpTypeArray %uchar %ulong_1
%_ptr_Function__arr_uchar_ulong_1 = OpTypePointer Function %_arr_uchar_ulong_1
%_ptr_CrossWorkgroup_uchar = OpTypePointer CrossWorkgroup %uchar
          %9 = OpTypeFunction %void %_ptr_Function__arr_uchar_ulong_1 %_ptr_CrossWorkgroup_uchar
%_ptr_Function_uchar = OpTypePointer Function %uchar
         %10 = OpFunction %void None %9
         %11 = OpFunctionParameter %_ptr_Function__arr_uchar_ulong_1
         %12 = OpFunctionParameter %_ptr_CrossWorkgroup_uchar
        %top = OpLabel
         %16 = OpInBoundsPtrAccessChain %_ptr_Function_uchar %11 %ulong_0 %ulong_0
         %17 = OpLoad %uchar %16 Aligned 1
               OpStore %12 %17 Aligned 1
               OpReturn
               OpFunctionEnd
```

The kernel is a very simple one, reading a byval i8 value and writing it to an output pointer. However, executing the above code does not behave as such, and the input value isn't written to the output pointer. The culprit seems to be the `byval([1 x i8])`, wrapping the value in a struct (which doesn't change the memory layout) results in correct execution.

```llvm
define spir_kernel void @_Z17julia_kernel_60905TupleI4Int8E7LLVMPtrIS0_Li1EE({ [1 x i8] }* byval({ [1 x i8] }) %0, i8 addrspace(1)* %1) local_unnamed_addr #0 {
conversion:
  %2 = getelementptr inbounds { [1 x i8] }, { [1 x i8] }* %0, i64 0, i32 0, i64 0
  %3 = load i8, i8* %2, align 1
  store i8 %3, i8 addrspace(1)* %1, align 1
  ret void
}
```

```
; SPIR-V
; Version: 1.0
; Generator: Khronos LLVM/SPIR-V Translator; 14
; Bound: 21
; Schema: 0
               OpCapability Addresses
               OpCapability Kernel
               OpCapability Int64
               OpCapability Int8
          %1 = OpExtInstImport "OpenCL.std"
               OpMemoryModel Physical64 OpenCL
               OpEntryPoint Kernel %11 "_Z17julia_kernel_61175TupleI4Int8E7LLVMPtrIS0_Li1EE"
               OpSource OpenCL_C 200000
               OpName %conversion "conversion"
               OpDecorate %12 FuncParamAttr ByVal
      %uchar = OpTypeInt 8 0
      %ulong = OpTypeInt 64 0
       %uint = OpTypeInt 32 0
    %ulong_1 = OpConstant %ulong 1
    %ulong_0 = OpConstant %ulong 0
     %uint_0 = OpConstant %uint 0
       %void = OpTypeVoid
%_arr_uchar_ulong_1 = OpTypeArray %uchar %ulong_1
  %_struct_3 = OpTypeStruct %_arr_uchar_ulong_1
%_ptr_Function__struct_3 = OpTypePointer Function %_struct_3
%_ptr_CrossWorkgroup_uchar = OpTypePointer CrossWorkgroup %uchar
         %10 = OpTypeFunction %void %_ptr_Function__struct_3 %_ptr_CrossWorkgroup_uchar
%_ptr_Function_uchar = OpTypePointer Function %uchar
         %11 = OpFunction %void None %10
         %12 = OpFunctionParameter %_ptr_Function__struct_3
         %13 = OpFunctionParameter %_ptr_CrossWorkgroup_uchar
 %conversion = OpLabel
         %19 = OpInBoundsPtrAccessChain %_ptr_Function_uchar %12 %ulong_0 %uint_0 %ulong_0
         %20 = OpLoad %uchar %19 Aligned 1
               OpStore %13 %20 Aligned 1
               OpReturn
               OpFunctionEnd
```

FWIW, the LLVM IR originates from the following Julia source code:

```julia
using oneAPI

function kernel(val, out)
    @inbounds unsafe_store!(out, val[1])
    return
end

val = (rand(Int8),)
d_out = oneArray(zeros(Int8))

@oneapi kernel(val, reinterpret(Core.LLVMPtr{Int8, 1}, pointer(d_out)))
```

I thought I had filed an issue about this before, but I can't find it.

**Comment by mnaczk** [comment]
According to the OneApi documentation https://spec.oneapi.io/level-zero/latest/core/SPIRV.html#kernel-arguments
The kernel argument cannot be OpTypeArray.

In the SpirV code submitted by you:
```
...
OpDecorate %12 FuncParamAttr ByVal
...
% _arr_uchar_ulong_1 = OpTypeArray% uchar% ulong_1
% _ptr_Function__arr_uchar_ulong_1 = OpTypePointer Function% _arr_uchar_ulong_1
…
% 11 = OpFunctionParameter% _ptr_Function__arr_uchar_ulong_1
...
```
It can be seen that the OpTypePointer type is passed to the OpFunctionParameter.
OpTypePointer is supported. However, the Function Parameter Attribute ByVal is also passed.

Based on this documentation: https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#_function_parameter_attribute

ByVal 'unpacks" this OpTypePointer and in a fact passes as the argument of the type pointed by the pointer. In your case, it is an OpTypeArray that is forbidden by the specification and for which we do not have support.

To sum up. By passing to the kernel as an argument the pointer with Function Parameter Attribute ByVal,
this pointer must not point to an OpTypeArray.  **We do not plan to support this type as a kernel argument in the nearest future.**

The solution to the problem is to 'wrap' an illegal type into a structure that is already a type supported by OneApi specifications.

**Comment by maleadt** [comment]
What's the reasoning for not supporting these, when they're allowed as struct members, and the memory layout is identical?

If unsupported, I'd argue that this should throw an error instead of resulting in silent miscompilation.

**Comment by mnaczk** [comment]
I close the issue.
We do not plan to support OpTypeArray as a kernel argument. The proper error message is implemented.
